cmd/compute-domain-daemon/dnsnames.go

```diff
@@ -0,0 +1,208 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"k8s.io/klog/v2"
+
+	nvapi "github.com/NVIDIA/k8s-dra-driver-gpu/api/nvidia.com/resource/v1beta1"
+)
+
+const (
+	hostsFilePath = "/etc/hosts"
+	dnsNameFormat = "compute-domain-daemon-%d"
+)
+
+// DNSNameManager manages the allocation of static DNS names to IP addresses.
+type DNSNameManager struct {
+	sync.Mutex
+	ipToDNSName           map[string]string
+	cliqueID              string
+	maxNodesPerIMEXDomain int
+	nodesConfigPath       string
+}
+
+// NewDNSNameManager creates a new DNS name manager.
+func NewDNSNameManager(cliqueID string, maxNodesPerIMEXDomain int, nodesConfigPath string) *DNSNameManager {
+	return &DNSNameManager{
+		ipToDNSName:           make(map[string]string),
+		cliqueID:              cliqueID,
+		maxNodesPerIMEXDomain: maxNodesPerIMEXDomain,
+		nodesConfigPath:       nodesConfigPath,
+	}
+}
+
+// UpdateDNSNameMappings updates the /etc/hosts file with IP to DNS name mappings.
+func (m *DNSNameManager) UpdateDNSNameMappings(nodes []*nvapi.ComputeDomainNode) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// Prefilter nodes to only consider those with the matching cliqueID
+	var cliqueNodes []*nvapi.ComputeDomainNode
+	for _, node := range nodes {
+		if node.CliqueID == m.cliqueID {
+			cliqueNodes = append(cliqueNodes, node)
+		}
+	}
+
+	// Find and remove stale IPs from map
+	currentIPs := make(map[string]bool)
+	for _, node := range cliqueNodes {
+		currentIPs[node.IPAddress] = true
+	}
+
+	for ip := range m.ipToDNSName {
+		if !currentIPs[ip] {
+			delete(m.ipToDNSName, ip)
+		}
+	}
+
+	// Add new IPs to map (filling in holes where others were removed)
+	for _, node := range cliqueNodes {
+		// If IP already has a DNS name, skip it
+		if _, exists := m.ipToDNSName[node.IPAddress]; exists {
+			continue
+		}
+
+		dnsName, err := m.allocateDNSName(node)
+		if err != nil {
+			return fmt.Errorf("failed to allocate DNS name for IP %s: %w", node.IPAddress, err)
+		}
+		m.ipToDNSName[node.IPAddress] = dnsName
+	}
+
+	// Update the hosts file with current mappings
+	return m.updateHostsFile()
+}
+
+// LogDNSNameMappings logs the current compute-domain-daemon mappings from memory.
+func (m *DNSNameManager) LogDNSNameMappings() {
+	m.Lock()
+	defer m.Unlock()
+
+	if len(m.ipToDNSName) == 0 {
+		klog.Infof("No compute-domain-daemon mappings found")
+		return
+	}
+
+	klog.Infof("Current compute-domain-daemon mappings:")
+	for ip, dnsName := range m.ipToDNSName {
+		klog.Infof("  %s -> %s", ip, dnsName)
```

Comment:
For further down the line: something like that should be part of a single log message, structurally.

Reply:
> I find it harder to read on a single line

In structured logging, a single log message has metadata attached, and the actual log message payload is a string which totally can include newline characters :).

In other words, only very bad log propagation systems do _not_ understand multi-line log messages.